### PR TITLE
KT-7930

### PR DESCRIPTION
**Pull request by voddan**
Added  extensions`String.toIntOrNull` etc; Optimased basing on openJDK implementations

### CODE REVIEW

**Comment by voddan** [comment]
I purpose resolving some of the functionality of https://youtrack.jetbrains.com/issue/KT-8286 together with this PR

One part of KT-8286 is providing `String.toInt()` with an optional parameter `String.toInt(radix=10)`. This is very easy to add in this PR since it affects those functions. More other, some functions in this PR (`String.toIntOrNull()` and `String.toLongOrNull()`) were not made `inline`, which would be an obstacle for adding an additional functional parameter to them later due to binary compatibility.

I can 
a) add a commit to this PR with the changes described above
b) create a new PR after that one is closed and implement the functionality there

**Comment by voddan** [comment]
this two commits add DSL for the tests, streamlining the code. I hope they don't add too much extra mental load

File: libraries/stdlib/test/text/ParsePrimitivesJVMTest.kt [hunk 1]
```diff
@@ -13,79 +13,96 @@ class ParsePrimitivesJVMTest {
...
-        assertNull (convertOrNull(input) )
-    } else {
-        assertEquals(output, converOrFail(input))
-        assertEquals(output, convertOrNull(input))
+
+private class CompareBehaviourContext<T: Any>(val convertOrFail: (String) -> T,
+                                              val convertOrNull: (String) -> T?) {
+    fun assertProduce(output: T?, input: String) {
+        if(output == null) {
+            assertFails { convertOrFail(input) }
```
Comment (ilya-g) [inline libraries/stdlib/test/text/ParsePrimitivesJVMTest.kt]: Please specify input string in the detail message, i.e. something like "Expected to fail (or return null) on $input"

**Comment by ilya-g** [comment]
> I hope they don't add too much extra mental load
Nevermind, I'll squash some commits anyway.
Btw, you can do it yourself and force push to your branch.

File: libraries/stdlib/test/text/ParsePrimitivesJVMTest.kt [hunk 2]
```diff
@@ -13,79 +13,96 @@ class ParsePrimitivesJVMTest {
...
-                                                        crossinline converOrFail: (String) -> T,
-                                                        crossinline convertOrNull: (String) -> T?) {
-    if(output == null) {
-        assertFails { converOrFail(input) }
-        assertNull (convertOrNull(input) )
-    } else {
-        assertEquals(output, converOrFail(input))
-        assertEquals(output, convertOrNull(input))
+
+private class CompareBehaviourContext<T: Any>(val convertOrFail: (String) -> T,
```
Comment (ilya-g) [inline libraries/stdlib/test/text/ParsePrimitivesJVMTest.kt]: `Behaviour` we prefer American English in identifiers :)

**Comment by voddan** [comment]
@ilya-g by mental load I meant the whole DSL thing.

**Comment by ilya-g** [comment]
Oh, then it's ok. It's compensated by much less mental load from clean tests.

**Comment by voddan** [comment]
@ilya-g What about the `radix` versions from KT-8286?

**Comment by ilya-g** [comment]
I propose making it a separate PR (branched on the top of this one)

**Comment by ilya-g** [comment]
This was superseded by PR #846

**Pull request closed without merging**